docs/source/developers/release_verification.rst

```diff
@@ -0,0 +1,148 @@
+non-binding votes are greatly encouraged and a sign of a healthy project.
+
+Running the release verification
+================================
+
+Linux and macOS
+---------------
+
+In order to run the verification script either for the source release or the
+binary artifacts see the following guidelines:
+
+.. code-block::
+
+   # this will create and automatically clean up a temporary directory for the verification environment and will run the source verification
+   TEST_DEFAULT=0 TEST_SOURCE=1 verify-release-candidate.sh $VERSION $RC_NUM
+   
+   # this will create and automatically clean up a temporary directory for the verification environment and will run the binary verification
+   TEST_DEFAULT=0 TEST_BINARIES=1 dev/release/verify-release-candidate.sh $VERSION $RC_NUM
+   
+   # to verify only certain implementations use the TEST_DEFAULT=0 and TEST_* variables
+   # here are a couple of examples, but see the source code for the available options
+   TEST_DEFAULT=0 TEST_CPP=1 verify-release-candidate.sh $VERSION $RC_NUM  # only C++ tests
+   TEST_DEFAULT=0 TEST_CPP=1 TEST_PYTHON=1 verify-release-candidate.sh $VERSION $RC_NUM  # C++ and Python tests
+   TEST_DEFAULT=0 TEST_INTEGRATION_CPP=1 TEST_INTEGRATION_JAVA=1 verify-release-candidate.sh $VERSION $RC_NUM  # C++ and Java integration tests
+   
+   # to verify certain binaries use the TEST_* variables as:
+   TEST_DEFAULT=0 TEST_WHEELS=1 verify-release-candidate.sh $VERSION $RC_NUM  # only Wheels
+   TEST_DEFAULT=0 TEST_APT=1 verify-release-candidate.sh $VERSION $RC_NUM  # only APT packages
+   TEST_DEFAULT=0 TEST_YUM=1 verify-release-candidate.sh $VERSION $RC_NUM  # only YUM packages
+   TEST_DEFAULT=0 TEST_JARS=1 verify-release-candidate.sh $VERSION $RC_NUM  # only JARS
+
+Windows
+-------
+
+In order to run the verification script on Windows you have to download
+the source tarball from the SVN dist system that you wish to verify:
+
+.. code-block::
+
+   dev\release\verify-release-candidate.bat %VERSION% %RC_NUM%
+
+System Configuration Instructions
+=================================
+
+You will need some tools installed like curl, git, etcetera.
+
+Ubuntu
+------
+
+You might have to install some packages on your system. The following
+utility script can be used to set your Ubuntu system. This wil install
+the required packages to perform a source verification on a clean
+Ubuntu:
+
+.. code-block::
+
+   # From the arrow clone
+   popd dev/release
+   sudo ./setup-ubuntu.sh
+   pushd
+
+macOS ARM
+---------
+
+.. code-block::
+
+   # From the arrow clone
+   brew install gpg
+   brew bundle --file=cpp/Brewfile
+   brew bundle --file=c_glib/Brewfile
+   brew uninstall node
+   # You might need to add node, ruby java and maven to the PATH, follow
+   # instructions from brew after installing.
+   brew install node@20
+   brew install ruby
+   brew install openjdk
+   brew install maven
+   # libffi required for gobject-introspection gem
+   brew install libffi
```

Comment:
I think that this is needless because Homebrew's GObject Introspection uses system libffi: https://github.com/Homebrew/homebrew-core/blob/c8900f1b3f0483b602e948d2c5400e8ac870e690/Formula/g/gobject-introspection.rb#L33

```suggestion
```

Reply:
without this my build was failing on a brand new M2. I can share the logs tomorrow if you want.

Reply:
Wow! Yes, please share the logs!

Reply:
After `brew uninstall libffi` this is the error I get:
```
Resolving dependencies...
Fetching power_assert 2.0.3
Fetching native-package-installer 1.1.9
Fetching pkg-config 1.5.6
Installing power_assert 2.0.3
Installing native-package-installer 1.1.9
Installing pkg-config 1.5.6
Fetching test-unit 3.6.2
Fetching glib2 4.2.0
Installing test-unit 3.6.2
Installing glib2 4.2.0 with native extensions
Fetching gobject-introspection 4.2.0
Installing gobject-introspection 4.2.0 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

current directory:
/private/var/folders/rr/t_15q7dx4_lf1nv0btypffcw0000gn/T/arrow-15.0.0.XXXXX.li0gqOMB3O/apache-arrow-15.0.0/c_glib/vendor/bundle/ruby/3.3.0/gems/gobject-introspection-4.2.0/ext/gobject-introspection
/opt/homebrew/opt/ruby/bin/ruby extconf.rb
checking for --enable-debug-build option... no
checking for -Wall option to compiler... yes
checking for -Wcast-align option to compiler... yes
checking for -Wextra option to compiler... yes
checking for -Wformat=2 option to compiler... yes
checking for -Winit-self option to compiler... yes
checking for -Wlarger-than-65500 option to compiler... yes
checking for -Wmissing-declarations option to compiler... yes
checking for -Wmissing-format-attribute option to compiler... yes
checking for -Wmissing-include-dirs option to compiler... yes
checking for -Wmissing-noreturn option to compiler... yes
checking for -Wmissing-prototypes option to compiler... yes
checking for -Wnested-externs option to compiler... yes
checking for -Wold-style-definition option to compiler... yes
checking for -Wpacked option to compiler... yes
checking for -Wp,-D_FORTIFY_SOURCE=2 option to compiler... yes
checking for -Wpointer-arith option to compiler... yes
checking for -Wundef option to compiler... yes
checking for -Wout-of-line-declaration option to compiler... yes
checking for -Wunsafe-loop-optimizations option to compiler... no
checking for -Wwrite-strings option to compiler... yes
checking for Homebrew... yes
checking for gobject-introspection-1.0... yes (1.78.1)
creating rbgiversion.h
creating gobject-introspection-enum-types.c
creating gobject-introspection-enum-types.h
creating ruby-gobject-introspection.pc
creating Makefile

current directory:
/private/var/folders/rr/t_15q7dx4_lf1nv0btypffcw0000gn/T/arrow-15.0.0.XXXXX.li0gqOMB3O/apache-arrow-15.0.0/c_glib/vendor/bundle/ruby/3.3.0/gems/gobject-introspection-4.2.0/ext/gobject-introspection
make DESTDIR\= sitearchdir\=./.gem.20240216-45788-684atu sitelibdir\=./.gem.20240216-45788-684atu clean

current directory:
/private/var/folders/rr/t_15q7dx4_lf1nv0btypffcw0000gn/T/arrow-15.0.0.XXXXX.li0gqOMB3O/apache-arrow-15.0.0/c_glib/vendor/bundle/ruby/3.3.0/gems/gobject-introspection-4.2.0/ext/gobject-introspection
make DESTDIR\= sitearchdir\=./.gem.20240216-45788-684atu sitelibdir\=./.gem.20240216-45788-684atu
compiling gobject-introspection-enum-types.c
compiling rb-gi-arg-info.c
In file included from rb-gi-arg-info.c:21:
In file included from ./rb-gi-private.h:24:
In file included from
/private/var/folders/rr/t_15q7dx4_lf1nv0btypffcw0000gn/T/arrow-15.0.0.XXXXX.li0gqOMB3O/apache-arrow-15.0.0/c_glib/vendor/bundle/ruby/3.3.0/gems/glib2-4.2.0/ext/glib2/../../lib/rbgobject.h:27:
In file included from
/private/var/folders/rr/t_15q7dx4_lf1nv0btypffcw0000gn/T/arrow-15.0.0.XXXXX.li0gqOMB3O/apache-arrow-15.0.0/c_glib/vendor/bundle/ruby/3.3.0/gems/glib2-4.2.0/ext/glib2/../../lib/rbgutil.h:26:
In file included from /opt/homebrew/Cellar/ruby/3.3.0/include/ruby-3.3.0/ruby/encoding.h:22:
/opt/homebrew/Cellar/ruby/3.3.0/include/ruby-3.3.0/ruby/internal/encoding/ctype.h:82:46: warning: unused parameter 'enc' [-Wunused-parameter]
rb_enc_isascii(OnigCodePoint c, rb_encoding *enc)
                                             ^
In file included from rb-gi-arg-info.c:21:
In file included from ./rb-gi-private.h:28:
/opt/homebrew/Cellar/gobject-introspection/1.78.1/include/gobject-introspection-1.0/girffi.h:25:10: fatal error: 'ffi.h' file not found
#include <ffi.h>
         ^~~~~~~
1 warning and 1 error generated.
make: *** [rb-gi-arg-info.o] Error 1

make failed, exit code 2

Gem files will remain installed in
/private/var/folders/rr/t_15q7dx4_lf1nv0btypffcw0000gn/T/arrow-15.0.0.XXXXX.li0gqOMB3O/apache-arrow-15.0.0/c_glib/vendor/bundle/ruby/3.3.0/gems/gobject-introspection-4.2.0 for inspection.
Results logged to
/private/var/folders/rr/t_15q7dx4_lf1nv0btypffcw0000gn/T/arrow-15.0.0.XXXXX.li0gqOMB3O/apache-arrow-15.0.0/c_glib/vendor/bundle/ruby/3.3.0/extensions/arm64-darwin-23/3.3.0/gobject-introspection-4.2.0/gem_make.out

  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/builder.rb:125:in `run'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/builder.rb:51:in `block in make'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/builder.rb:43:in `each'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/builder.rb:43:in `make'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/ext_conf_builder.rb:42:in `build'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/builder.rb:193:in `build_extension'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/builder.rb:227:in `block in build_extensions'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/builder.rb:224:in `each'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/ext/builder.rb:224:in `build_extensions'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/installer.rb:852:in `build_extensions'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/rubygems_gem_installer.rb:76:in `build_extensions'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/rubygems_gem_installer.rb:28:in `install'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/source/rubygems.rb:205:in `install'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/installer/gem_installer.rb:54:in `install'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/installer/gem_installer.rb:16:in `install_from_spec'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/installer/parallel_installer.rb:132:in `do_install'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/installer/parallel_installer.rb:123:in `block in worker_pool'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/worker.rb:62:in `apply_func'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/worker.rb:57:in `block in process_queue'
  <internal:kernel>:187:in `loop'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/worker.rb:54:in `process_queue'
  /opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/bundler/worker.rb:90:in `block (2 levels) in create_threads'

An error occurred while installing gobject-introspection (4.2.0), and Bundler cannot continue.

In Gemfile:
  gobject-introspection
Failed to verify release candidate. See /var/folders/rr/t_15q7dx4_lf1nv0btypffcw0000gn/T/arrow-15.0.0.XXXXX.li0gqOMB3O for details.
```

Reply:
Thanks!

Could you show the output of `pkg-config --cflags gobject-introspection-1.0` on the environment?

Reply:
sure:
```
% pkg-config --cflags gobject-introspection-1.0
-I/opt/homebrew/Cellar/gobject-introspection/1.78.1/include/gobject-introspection-1.0 -I/opt/homebrew/Cellar/glib/2.78.4/include -I/opt/homebrew/Cellar/glib/2.78.4/include/glib-2.0 -I/opt/homebrew/Cellar/glib/2.78.4/lib/glib-2.0/include -I/opt/homebrew/opt/gettext/include -I/opt/homebrew/Cellar/pcre2/10.42/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/ffi
```

Reply:
Thanks!

It includes `-I/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/ffi`. It seems that the environment doesn't have `/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/`.

Could you show the output of `ls -l /Library/Developer/CommandLineTools/SDKs/`?

But it's a bit strange. I think that a brand new M2 uses more recent version than 12... Could you also show the output of `sw_vers -productVersion`? Homebrew uses it to detect the version.

Reply:
Thanks. `MacOSX12.sdk -> MacOSX12.3.sdk` symbolic link doesn't exist. I think that it's a problem.

Reply:
On my Mac mini environment:

```console
$ ls -l /Library/Developer/CommandLineTools/SDKs
total 0
lrwxr-xr-x  1 root  wheel   14 12  2 05:05 MacOSX.sdk -> MacOSX14.0.sdk
drwxr-xr-x  7 root  wheel  224  8 21  2023 MacOSX12.3.sdk
lrwxr-xr-x  1 root  wheel   14  8 21  2023 MacOSX12.sdk -> MacOSX12.3.sdk
drwxr-xr-x  7 root  wheel  224 12  2 05:05 MacOSX13.3.sdk
lrwxr-xr-x  1 root  wheel   14 12  2 05:04 MacOSX13.sdk -> MacOSX13.3.sdk
drwxr-xr-x  7 root  wheel  224  9  1 04:12 MacOSX14.0.sdk
lrwxr-xr-x  1 root  wheel   14 12  2 05:04 MacOSX14.sdk -> MacOSX14.0.sdk
```

Could you try again after `sudo ln -s MacOSX12.3.sdk /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk`?

Reply:
(I'm not sure who created the `MacOSX${MAJOR_VERSION}.sdk` symbolic links. But I haven't done it manually.)

Reply:
@kou as per the PR, if you want I can remove the `brew install libffi` and we can keep investigating further the issue on a different GH issue so we can merge this initial documentation, are you ok with that?

Reply:
Yes. Let's do it! (We remove `brew install libffi` in this PR, merge this PR and open a new issue for this.)

Reply:
Done and created issue to follow up:
* https://github.com/apache/arrow/issues/40195

Reply:
Thanks!